pkg/alertmanager/distributor/distributor.go

```diff
@@ -0,0 +1,283 @@
+// TODO: This will go away in future after we gradually add support for the entire API.
+func (d *Distributor) IsPathSupported(path string) bool {
+	// API can be found at https://petstore.swagger.io/?url=https://raw.githubusercontent.com/prometheus/alertmanager/master/api/v2/openapi.yaml.
+	return strings.HasSuffix(path, "/alerts") ||
+		strings.HasSuffix(path, "/alerts/groups")
+}
+
+// DistributeRequest shards the writes and returns as soon as the quorum is satisfied.
+// In case of reads, it proxies the request to one of the alertmanagers.
+// DistributeRequest assumes that the caller has verified IsPathSupported return
+// true for the route.
+func (d *Distributor) DistributeRequest(w http.ResponseWriter, r *http.Request) {
+	//TODO: Initialise tracing here.
+
+	d.requestsInFlight.Add(1)
+	defer d.requestsInFlight.Done()
+
+	userID, err := tenant.TenantID(r.Context())
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusUnauthorized)
+		return
+	}
+
+	d.receivedRequests.WithLabelValues(userID).Inc()
+
+	if r.Method == http.MethodGet {
+		d.doRead(userID, w, r)
+	} else {
+		d.doWrite(userID, w, r)
+	}
+}
+
+func (d *Distributor) doWrite(userID string, w http.ResponseWriter, r *http.Request) {
+	source := util.GetSourceIPsFromOutgoingCtx(r.Context())
+
+	var body []byte
+	var err error
+	if r.Body != nil {
+		body, err = ioutil.ReadAll(r.Body)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+		}
+	}
+
+	err = ring.DoBatch(r.Context(), ring.Write, d.alertmanagerRing, []uint32{shardByUser(userID)}, func(am ring.IngesterDesc, _ []int) error {
+		d.amSends.WithLabelValues(am.Addr).Inc()
+
+		// Use a background context to make sure all alertmanagers get the request even if we return early.
+		localCtx, cancel := context.WithTimeout(context.Background(), d.cfg.RemoteTimeout)
+		defer cancel()
+		localCtx = user.InjectOrgID(localCtx, userID)
+		if sp := opentracing.SpanFromContext(r.Context()); sp != nil {
+			localCtx = opentracing.ContextWithSpan(localCtx, sp)
+		}
+		// Get clientIP(s) from Context and add it to localCtx
+		localCtx = util.AddSourceIPsToOutgoingContext(localCtx, source)
+
+		// http.Request.Clone() does not do a deep copy, hence copying the body.
+		// https://github.com/golang/go/issues/36095.
+		rClone := r.Clone(r.Context())
+		rClone.Body = ioutil.NopCloser(bytes.NewReader(body))
+
+		resp, err := d.doRequest(localCtx, am, rClone)
+		if err != nil {
+			d.amSendFailures.WithLabelValues(am.Addr).Inc()
+			return err
+		}
+
+		if resp.GetStatus() != alertmanagerpb.OK {
+			return errors.New("alertmanager grpc request not ok")
+		}
+
+		return nil
+	}, func() {})
+
+	if err == nil {
+		w.WriteHeader(http.StatusOK)
+		return
+	}
+
+	d.respondFromError(err, w)
+}
+
+func (d *Distributor) doRead(userID string, w http.ResponseWriter, r *http.Request) {
+	key := shardByUser(userID)
+	replicationSet, err := d.alertmanagerRing.Get(key, ring.Read, nil, nil, nil)
```

Comment:
```suggestion
	replicationSet, err := d.alertmanagerRing.Get(key, alertmanager.RingOp, nil, nil, nil)
```

Ditto from above, given the current semantics I don't think we want to read from anything that's not active.